### PR TITLE
Assign indices of PIT requests in transport layer

### LINKED ISSUE #84187: Move point in time setup to the transport layer?
Some parts of the point in time setup happen on the REST layer, see https://github.com/elastic/elasticsearch/blob/13390dd4d820834a4b0135bfb5d24618accb15f2/server/src/main/java/org/elasticsearch/rest/action/search/RestSearchAction.java#L373-L374 . This causes problems to consumers that call search directly at the transport layer. Either those consumers reploicate the same setup on their side before calling search, or that setup is shared and moved from the REST layer to `TransportSearchAction`?

In practice, this ends up causing EQL to hit all indices when using point in time, instead of the configured indices.

### PR DESCRIPTION
**Pull request by dnhatn**
- Reject search requests that have both indices and point-in-time
- Assign the indices of PIT requests in the transport layer; otherwise, the indices of internal requests aren't assigned correctly

Closes #84187

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/action/support/TransportAction.java [hunk 1]
```diff
@@ -54,6 +56,13 @@ public final void execute(Task task, Request request, ActionListener<Response> l
         requestFilterChain.proceed(task, actionName, request, listener);
     }
 
+    /**
+     * Prepare the request before executing it. Defaults to no-op.
+     */
+    protected void prepareRequest(Request request) {
```
Comment (dnhatn) [inline server/src/main/java/org/elasticsearch/action/support/TransportAction.java:62]: I tried to make this an ActionFilter, but we need a hack to do this. The new action filter must be executed before the security action filter, but the security action filter already has the lowest order (Integer.MIN_VALUE).

Comment (javanna) [reply server/src/main/java/org/elasticsearch/action/support/TransportAction.java:62]: I see the problem Nhat, thanks for raising this. Adding a new base method that runs before action filters makes me nervous. Possibly an alternative hack: would it work to make what SearchReqest#indices returns dynamic and extract the indices from pit if a pit is set? In that case the authorization filter would look at the right indices and we would not need to set the indices manually? I guess there can't be wildcards in there hence the authorization filter will not replace the indices in that scenario. there may be other side-effects though around going down this route.

File: server/src/internalClusterTest/java/org/elasticsearch/search/slice/SearchSliceIT.java [hunk 2]
```diff
@@ -226,10 +226,11 @@ private void assertSearchSlicesWithPointInTime(String sliceField, String sortFie
         for (int id = 0; id < numSlice; id++) {
             int numSliceResults = 0;
 
-            SearchRequestBuilder request = client().prepareSearch("test")
+            SearchRequestBuilder request = client().prepareSearch()
                 .slice(new SliceBuilder(sliceField, id, numSlice))
                 .setPointInTime(new PointInTimeBuilder(pointInTimeId))
                 .addSort(SortBuilders.fieldSort(sortField))
+                .setPreference(null)
```
Comment (javanna) [inline server/src/internalClusterTest/java/org/elasticsearch/search/slice/SearchSliceIT.java:233]: out of curiosity, why set the preference to null explicitly?

File: server/src/main/java/org/elasticsearch/action/search/SearchRequest.java [hunk 3]
```diff
@@ -345,6 +345,15 @@ public ActionRequestValidationException validate() {
...
             }
+            if (indices.length > 0) {
+                validationException = addValidationError("[indices] cannot be used with point in time", validationException);
+            }
+            if (routing != null) {
+                validationException = addValidationError("[routing] cannot be used with point in time", validationException);
+            }
+            if (preference != null) {
+                validationException = addValidationError("[preference] cannot be used with point in time", validationException);
+            }
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/action/search/SearchRequest.java:356]: Thanks, ++ on having these checks as part of the request validation

File: server/src/main/java/org/elasticsearch/action/support/TransportAction.java (hunk 1 above)
Comment (javanna) [reply server/src/main/java/org/elasticsearch/action/support/TransportAction.java:62]: I see the problem Nhat, thanks for raising this. Adding a new base method that runs before action filters makes me nervous. Possibly an alternative hack: would it work to make what SearchReqest#indices returns dynamic and extract the indices from pit if a pit is set? In that case the authorization filter would look at the right indices and we would not need to set the indices manually? I guess there can't be wildcards in there hence the authorization filter will not replace the indices in that scenario. there may be other side-effects though around going down this route.

File: server/src/main/java/org/elasticsearch/rest/action/search/RestMultiSearchAction.java [hunk 4]
```diff
@@ -137,7 +138,12 @@ public static MultiSearchRequest parseRequest(
             searchRequest.source(SearchSourceBuilder.fromXContent(parser, false));
             RestSearchAction.checkRestTotalHits(restRequest, searchRequest);
             if (searchRequest.pointInTimeBuilder() != null) {
-                RestSearchAction.preparePointInTime(searchRequest, restRequest, namedWriteableRegistry);
+                if (restRequest.paramAsBoolean("ccs_minimize_roundtrips", false)) {
+                    throw new IllegalArgumentException("[ccs_minimize_roundtrips] cannot be used with point in time");
+                }
+                if (searchRequest.indicesOptions().equals(DEFAULT_INDICES_OPTIONS) == false) {
+                    throw new IllegalArgumentException("[indicesOptions] cannot be used with point in time");
+                }
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/rest/action/search/RestMultiSearchAction.java:146]: I wonder why we need these at the REST layer, shouldn't these checks be performed as part of the request validation anyways. Would we fail anyways down the line without these additional checks?

File: server/src/main/java/org/elasticsearch/rest/action/search/RestSearchAction.java [hunk 5]
```diff
@@ -201,7 +197,12 @@ public static void parseSearchRequest(
         checkRestTotalHits(request, searchRequest);
 
         if (searchRequest.pointInTimeBuilder() != null) {
-            preparePointInTime(searchRequest, request, namedWriteableRegistry);
+            if (request.paramAsBoolean("ccs_minimize_roundtrips", false)) {
+                throw new IllegalArgumentException("[ccs_minimize_roundtrips] cannot be used with point in time");
+            }
+            if (searchRequest.indicesOptions().equals(DEFAULT_INDICES_OPTIONS) == false) {
+                throw new IllegalArgumentException("[indicesOptions] cannot be used with point in time");
+            }
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/rest/action/search/RestSearchAction.java:205]: Along the same lines as above, I wonder if we should try and share the validation that we already have in SearchRequest

File: x-pack/plugin/async-search/qa/security/src/javaRestTest/java/org/elasticsearch/xpack/search/AsyncSearchSecurityIT.java [hunk 6]
```diff
@@ -204,7 +204,7 @@ public void testRejectPointInTimeWithIndices() throws Exception {
         try {
             final Request request = new Request("POST", "/_async_search");
             setRunAsHeader(request, authorizedUser);
-            request.addParameter("wait_for_completion_timeout", "true");
+            request.addParameter("wait_for_completion_timeout", "10m");
```
Comment (javanna) [inline x-pack/plugin/async-search/qa/security/src/javaRestTest/java/org/elasticsearch/xpack/search/AsyncSearchSecurityIT.java:207]: interesting, this was a mistake that was not though causing any failure?

File: server/src/main/java/org/elasticsearch/action/support/TransportAction.java (hunk 1 above)
Comment (javanna) [reply server/src/main/java/org/elasticsearch/action/support/TransportAction.java:62]: @ywangd what do you think?

**Review by javanna — COMMENTED** [review]
I left some comments for discussions

File: server/src/main/java/org/elasticsearch/action/support/TransportAction.java (hunk 1 above)
Comment (ywangd) [reply server/src/main/java/org/elasticsearch/action/support/TransportAction.java:62]: If we make `SearchRequest#indices()` method dynamic, how would the corresponding setter behave? `SearchRequest` implements `IndicesRequest.Replaceable`. The contract for authorization code is that it will always call `SearchRequest#indices(String ..)` with the expanded indices regardless whether the expanded indices is identical to the original one. 

IIUC, in practice, Search with PIT always gets back to the identical list of indices from Authorization because it does not allow `ignore_unavailable=true`. So technically the setter can be a no-op, but I am not really comfortable with that because it feels PIT has to know how Authorization works. It also excludes the possibility to ever support `ignore_unavailable=true`. One option is to make the getter dynamically return the list indices from PIT if setter is never called, but otherwise return the value configured by the setter method (maybe this is already what @javanna suggests by "dynamic"?). The downside is inconsistency between the `indices()` method and the internal `indices` variable (till the setter is called).

I am ok with what @dnhatn has proposed. Though I'd rather add the new base method to `ActionRequest` instead of `TransportAction` because calling `ActionRequest#validate()` and `ActionRequest#prepare()` in sequence seems to be more fluent to me (it would be nice to make `ActionRequest#prepare` idempotent). 

Overall, I slightly prefer the `TransportAction` (or `ActionRequest`) approach which feels simpler.

Comment (javanna) [reply server/src/main/java/org/elasticsearch/action/support/TransportAction.java:62]: as far as I understand the pit indices never hold wildcards, and never will, because you need to first open a pit on a fixed set of indices. So the whole indices expansion and replacement does not really apply to this specific scenario. It would be great to have a different request for this scenario, but we don't unfortunately. The indices variable is only read by methods within SearchRequest that could call the indices() method instead for consistency? We could even say that the setter throws an exception in case a pit is set.

Comment (ywangd) [reply server/src/main/java/org/elasticsearch/action/support/TransportAction.java:62]: Indices expansion in authorization does not have knowledge about PIT. Its behaviour is largely determined by whether the incoming request is a `IndicesRequest.Replaceable`. If we could turn SearchRequest with PIT into a new class that does not implementing this interface, things should just work on the authorization side. I think this would be better than throwing exception in the setter method.

Comment (javanna) [reply server/src/main/java/org/elasticsearch/action/support/TransportAction.java:62]: > Indices expansion in authorization does not have knowledge about PIT.

Correct. What I am saying is that because you need to explicitly open a PIT with a specific API call, the PIT id will always hold concrete indices and never wildcards when set to a search request, because you previously received it as part of the open pit response, so despite the search request implements IndicesRequest.Replaceable, there will never be anything to expand when a PIT id is set. You were reasoning about ignore_unavailable and wildcards expansion in your previous reply, but these don't apply to a search when a PIT id is set to it. I think this is important as it reduces the scope of how indices are resolved and should be authorized in this specific scenario. I believe this is key to finding a simple solution for the current problem. I do not think that introducing a new base method to TransportAction for this very specific edge-case is a good trade-off.

> If we could turn SearchRequest with PIT into a new class that does not implementing this interface, things should just work on the authorization side. I think this would be better than throwing exception in the setter method.

That would make things better, and maybe it is possible given that we have control over the REST action, though it looks to me as it has quite a cost in terms of added complexity and lines of code. We would need to try that to see it in reality.

Comment (ywangd) [reply server/src/main/java/org/elasticsearch/action/support/TransportAction.java:62]: > though it looks to me as it has quite a cost in terms of added complexity and lines of code.

Yeah Java does not make delegation easy. 

> You were reasoning about ignore_unavailable and wildcards expansion in your previous reply, but these don't apply to a search when a PIT id is set to it.

Right I understand PIT does not really need expansion. My previous reply unnecessarily blurred the focus. My point is Authorization code will call `Replaceable#indices(String...)` as long as the request is an instance of `Replaceable`. I don't particularly like having a special setter method for `SearchRequest` because it indicates SearchRequest knows about authorization. But I am ok with the opposite, i.e. Authorization knows more about SearchRequest because it already knows quite a bit about the requests. The following is where Authorization decides whether to expand wildcards (and in turn calls the setter method):

https://github.com/elastic/elasticsearch/blob/11599e1e5f269679f3b96da76760ebd90610ab1e/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/IndicesAndAliasesResolver.java#L131-L141

We can add an addition check for PIT before `if (indicesRequest instanceof IndicesRequest.Replaceable)` such as:
```java
if (indicesRequest instanceof final SearchRequest searchRequest && searchRequest.pointInTimeBuilder() != null) {
    return false;
}
```

The above check in conjunction with a dynamic `SearchRequest#indices()` getter method should work and the change is rather minimal. What do you think?

Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/action/support/TransportAction.java:62]: @javanna @ywangd Thanks for the discussion.

> The above check in conjunction with a dynamic SearchRequest#indices() getter method should work and the change is rather minimal. What do you think?

This works for me.

File: server/src/main/java/org/elasticsearch/action/search/SearchRequest.java [hunk 7]
```diff
@@ -556,7 +581,12 @@ public SearchType searchType() {
      */
     @Override
     public String[] indices() {
-        return indices;
+        final PointInTimeBuilder pit = pointInTimeBuilder();
+        if (pit != null) {
+            return pit.getActualIndices();
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/action/search/SearchRequest.java:586]: shall we assert here that indices is empty?

File: server/src/main/java/org/elasticsearch/action/search/SearchRequest.java [hunk 8]
```diff
@@ -415,12 +427,22 @@ public Version minCompatibleShardNode() {
...
+    }
+
     /**
      * Sets the indices the search will be executed on.
      */
     @Override
     public SearchRequest indices(String... indices) {
         validateIndices(indices);
+        if (pointInTimeBuilder() != null && indices.length > 0) {
+            throw new IllegalArgumentException("[indices] cannot be used with point in time");
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/action/search/SearchRequest.java:444]: I would rephrase to sasy "indices cannot be set when a point in time is provided" I wonder if this should rather be an illegal state exception

File: server/src/main/java/org/elasticsearch/rest/action/search/RestSearchAction.java [hunk 9]
```diff
@@ -196,12 +191,17 @@ public static void parseSearchRequest(
         }
         searchRequest.routing(request.param("routing"));
         searchRequest.preference(request.param("preference"));
-        searchRequest.indicesOptions(IndicesOptions.fromRequest(request, searchRequest.indicesOptions()));
+        final IndicesOptions indicesOptions = IndicesOptions.fromRequest(request, searchRequest.indicesOptions());
+        if (searchRequest.indicesOptions().equals(indicesOptions) == false) {
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/rest/action/search/RestSearchAction.java:195]: is this new conditional necessary? What if we set them anyway?

File: x-pack/plugin/frozen-indices/src/internalClusterTest/java/org/elasticsearch/index/engine/frozen/FrozenIndexTests.java [hunk 10]
```diff
@@ -146,12 +146,7 @@ public void testCloseFreezeAndOpen() throws Exception {
         String pitId = openReaders(TimeValue.timeValueMinutes(1), indexName);
         try {
             for (int from = 0; from < 3; from++) {
-                searchResponse = client().prepareSearch()
-                    .setIndicesOptions(IndicesOptions.STRICT_EXPAND_OPEN_FORBID_CLOSED)
```
Comment (javanna) [inline x-pack/plugin/frozen-indices/src/internalClusterTest/java/org/elasticsearch/index/engine/frozen/FrozenIndexTests.java:150]: what is the effect of no longer setting the indices options here? They had no effect anyways?

File: x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/ClientTransformIndexer.java [hunk 11]
```diff
@@ -433,6 +435,7 @@ private void injectPointInTimeIfNeeded(
             ActionListener.wrap(response -> {
                 PointInTimeBuilder newPit = new PointInTimeBuilder(response.getPointInTimeId()).setKeepAlive(PIT_KEEP_ALIVE);
                 namedPits.put(namedSearchRequest.v1(), newPit);
+                searchRequest.indices(Strings.EMPTY_ARRAY);
```
Comment (javanna) [inline x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/ClientTransformIndexer.java:438]: do we need to set empty indices here now?

**Review by javanna — COMMENTED** [review]
I left some comments, I had left other comments previously also for which I haven't seen a reply, and I have not repeated those with this round.

**Comment by dnhatn** [comment]
@javanna @ywangd 

>  Possibly an alternative hack: would it work to make what SearchReqest#indices return dynamic and extract the indices from pit if a pit is set?

While working on this proposal, I found it's fragile. We need to carefully set and unset the point-in-time when updating the indices. Below is one of our usages that can be an issue. If we don't unset the `point-in-time` before retrieving `indices()` then we can add extra index filters to the search request although the actual indices can be empty.

https://github.com/elastic/elasticsearch/blob/273eeddc14601988ad0a70f0b199d163c6e09f1d/x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/search/PITAwareQueryClient.java#L105-L110

> Though I'd rather add the new base method to ActionRequestinstead of TransportAction because calling ActionRequest#validate() and ActionRequest#prepare() in sequence seems to be more fluent to me (it would be nice to make ActionRequest#prepare idempotent).

Now I think we should go with Yang's proposal.

**Comment by ywangd** [comment]
> While working on this proposal, I found it's fragile. We need to carefully set and unset the point-in-time when updating the indices.

IIUC, the issue is that changing `point-in-time` for a `SearchRequest` also leads to change of the output from `SearchRequest.indices()`. This interaction could be unexpected and depending on the order of calls, the resulted `indices` can be different. If you want to use the `indices` subsequently in other places, it might not do what you meant to do.

If we could prevent setting `pointInTimeBuilder` when the `SearchRequest` has non-empty `indices`, would that help? In theory, it makes sense to prevent it from happening since we are already preventing it at the REST (API) layer. That said, I am aware it might be technically challenging since `SearchSourceBuider` does _not_ know about `SearchRequest`?

**Comment by javanna** [comment]
Nhat and I have discussed this and I was hoping that we could find a simpler, less intrusive solution for this problem. Given that the transport client is no longer around, we should not design our system for direct access to the transport layer, hence I would opt for adapting the search request setup on the client side to set the indices in addition to the point in time.

**Comment by javanna** [comment]
We decided not to proceed with this change, closing.
